src/Compilers/VisualBasic/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncMethodToClassRewriter.Expressions.vb

```diff
@@ -95,19 +96,89 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                            End If
+
+                            Dim conditionalReceiver = TryCast(node.ReceiverOpt, BoundConditionalAccessReceiverPlaceholder)
+
+                            If conditionalReceiver IsNot Nothing Then
+
+                                If _conditionalAccessReceiverPlaceholderReplacementInfo Is Nothing OrElse
+                                   _conditionalAccessReceiverPlaceholderReplacementInfo.PlaceholderId <> conditionalReceiver.PlaceholderId Then
+                                    Throw ExceptionUtilities.Unreachable
+                                End If
+
+                                Debug.Assert(_conditionalAccessReceiverPlaceholderReplacementInfo.IsSpilled)
+                                Debug.Assert(Not _conditionalAccessReceiverPlaceholderReplacementInfo.ForceCaptureIfReferenceType)
+                                _conditionalAccessReceiverPlaceholderReplacementInfo.ForceCaptureIfReferenceType = True
+
+                            ElseIf Not receiverType.IsReferenceType Then
+                                ' A case where T Is actually a class must be handled specially.
+                                ' Taking a reference to a class instance Is fragile because the value behind the 
+                                ' reference might change while arguments are evaluated. However, the call should be
+                                ' performed on the instance that is behind reference at the time we push the
+                                ' reference to the stack. So, for a class we need to emit a reference to a temporary
+                                ' location, rather than to the original location
 
-                Dim result = SpillExpressionsWithReceiver(receiverOpt, isReceiverOfAMethodCall:=True, expressions:=arguments, spillBuilder:=builder)
+                                Dim referenceReceiverBuilder As New SpillBuilder()
+                                Dim spilledReferenceReceiver As BoundExpression = SpillRValue(receiverOpt.MakeRValue(), referenceReceiverBuilder)
+                                spilledReferenceReceiver = referenceReceiverBuilder.BuildSequenceAndFree(Me.F, spilledReferenceReceiver)
+                                Dim referenceReceiverSpillSequence = TryCast(spilledReferenceReceiver, BoundSpillSequence)
+
+                                If referenceReceiverSpillSequence IsNot Nothing Then
+                                    ' If condition `(object)default(T) != null` is true at execution time,
+                                    ' the T Is a value type. And it is a reference type otherwise.
+                                    Dim isValueTypeCheck = Me.F.ReferenceIsNotNothing(Me.F.DirectCast(Me.F.DirectCast(Me.F.Null(), receiverType),
+                                                                                      Me.F.SpecialType(SpecialType.System_Object)))
```

Comment:
> Looks like this condition will be false for nullable value types.

Yes, but we can leave with that and make a copy of a nullable value type because today there is no other more accurate check that is going to be reliably recognized and removed along with the branch known to be false by JIT during specialization. Throughout both compilers we do not go into these details in comments to not "muddle the water".